### ISSUE TITLE
[MNG-5366] [Regression] resolveAlways does not force dependency resolution in Maven 3.0.4

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#6559]
**[Paul Gier](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pgier)** opened **[MNG-5366](https://issues.apache.org/jira/browse/MNG-5366?redirect=false)** and commented

Using Maven 3.0.4, artifacts can only be resolved a single time during the build lifecycle using the Maven 2.x dependency resolution API.  Using resolver.resolveAlways() should force re-resolution of the artifact, however if the artifact was already resolved once during the build, then it will not be re-resolved even when calling resolveAlways().

This works as expected in Maven 3.0.0-3.0.3, and the artifact is re-resolved.

---

**Affects:** 3.0.4, 3.0.5, 3.1.0, 3.1.1, 3.2.1, 3.2.2

**Issue Links:**
- [MDEP-405](https://issues.apache.org/jira/browse/MDEP-405) purge-local-repository on multi-module project results in failure

- [MNG-5562](https://issues.apache.org/jira/browse/MNG-5562) Upgrade Aether 1.0 when available
 (_**"depends upon"**_)

0 votes, 9 watchers

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

I've searched for the code which is responsible for the `"Skipped remote update check ..."`-message, and it is.... `org.eclipse.aether.internal.impl.DefaultUpdateCheckManager`

Based on its checks I think we can fix it like this:

```
public void resolveAlways( Artifact artifact, List<ArtifactRepository> remoteRepositories,
                           ArtifactRepository localRepository )
    throws ArtifactResolutionException, ArtifactNotFoundException
{
    DefaultRepositorySystemSession session = new DefaultRepositorySystemSession( getSession( localRepository ) );
    session.setUpdatePolicy( RepositoryPolicy.UPDATE_POLICY_ALWAYS );
    resolve( artifact, remoteRepositories, session );
}
```

Still need to write a test for it, though.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

```
if ( isAlreadyUpdated( session.getData(), updateKey ) )
        {
            if ( logger.isDebugEnabled() )
            {
                logger.debug( "Skipped remote update check for " + check.getItem()
                    + ", already updated during this session." );
            }

            check.setRequired( false );
            if ( error != null )
            {
                check.setException( newException( error, artifact, repository ) );
            }
        }
```

The code above is part of Aether. As you can see it doesn't respect the policy. It simply checks if the `session.getData()` contains the `updateKey`. This can't be fixed from the Maven code, so this needs to be fixes at Aether.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

I've created https://bugs.eclipse.org/bugs/show_bug.cgi?id=425825

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

This has been fixed in Aether and when the Aether 1.0 is released we'll absorb the fix.

**Comment by jira-importer** [comment]
**[Igor Fedorenko](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=igorf)** commented

This maybe fixed by the new aether I just integrated in maven, but hard to tell for sure without example project or unit test to demonstrate the original problem.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

`maven-dependency-plugin\src\it\projects\purge-local-repository-reresolve` should verify this.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed in Aether 0.9.0.M3 and updated in Maven with 2909b5a329fd946796428a8e7f46034bc3bcbd2f by Igor Fedorenko.
